ncclient/operations/third_party/iosxe/rpc.py

```diff
@@ -1,9 +1,500 @@
+        return self._event_time
+
+    @property
+    def subscription_id(self):
+        if not self._parsed:
+            self.parse()
+        return self._subscription_id
+
+    @property
+    def type(self):
+        if not self._parsed:
+            self.parse()
+        return self._type
+
+    @property
+    def datastore_ele(self):
+        if not self._parsed:
+            self.parse()
+        return self._datastore
+
+    @property
+    def datastore_xml(self):
+        if not self._parsed:
+            self.parse()
+        return etree.tostring(self._datastore)
+
+    @property
+    def root_ele(self):
+        if not self._parsed:
+            self.parse()
+        return self._root
+
+    @property
+    def root_xml(self):
+        if not self._parsed:
+            self.parse()
+        return etree.tostring(self._root)
+
+
+class YangPushListener(SessionListener):
+
+    """Class extending :class:`Session` listeners, which are notified when
+    a new RFC 5277 notification is received or an error occurs. Only a
+    single instance of this class should be added to the listeners
+    list, and then individual subscription callback should be added to
+    the single listener.
+
+    """
+    def __init__(self):
+        """Called by EstablishSubscription when a new NotificationListener is
+        added to a session.  used to keep track of connection and
+        subscription info in case connection gets dropped.
+        """
+        self.subscription_listeners = {}
+
+        
+    def add_subscription_listener(self, id, user_callback, user_errback):
+        self.subscription_listeners[id] = (user_callback, user_errback)
+
+        
+    def rekey_subscription_listener(self, old_id, new_id):
+        self.subscription_listeners[new_id] = self.subscription_listeners.pop(old_id)
+
+        
+    def remove_subscription_listener(self, id):
+        self.subscription_listeners.pop(id)
+
+        
+    def callback(self, root, raw):
+        """Called when a new RFC 5277 notification is received.
+
+        The *root* argument allows the callback to determine whether
+        the message is a notification.  Here, *root* is a tuple of
+        *(tag, attributes)* where *tag* is the qualified name of the
+        root element and *attributes* is a dictionary of its
+        attributes (also qualified names).  *raw* will contain the xml
+        notification as a string.
+        """
+        tag, attrs = root
+        if tag != qualify("notification", NETCONF_NOTIFICATION_NS):
+            # we just ignore any message not a notification
+            return
+
+        notif = YangPushNotification(raw)
+        if notif.invalid:
+            logger.error("Couldn't parse notification")
+            return
+        try:
+            user_callback, _ = self.subscription_listeners[notif.subscription_id]
+            user_callback(notif)
+        except:
+            logger.error("No callback for subscription_id=%d" % notif.subscription_id)
```

Comment:
Personally I'm not a big fan of catch-all `except` cases, as they can catch exceptions that weren't what's intended. Perhaps instead:

```python
if notif.subscription_id in self.subscription_listeners:
    user_callback, _ = self.subscription_listeners[notif.subscription_id]
    user_callback(notif)
else:
    logger.error("No callback for subscription_id=%d", notif.subscription_id)
```